### PR TITLE
evidence: refactor package

### PR DESCRIPTION
**Pull request by cmwaters**
Closes: #4675

## Description

This PR is aimed at addressing a collection of small issues related to the evidence package:

- [x] Create a doc.go file to better explain the process of the module
- [x] If evidence is expired before it is committed we should remove it from the db (reactor.go:200)
- [x] ~Use store as an interface~, remove store entirely and keep db operations with db and the main logic with pool.go (it is unnecessary to have all three)
- [x] Committed evidence doesn't need to be stored completely, rather a marker pointing to the block in the block store (i.e. the height it is committed and the hash of the evidence for easy lookup)
- [x] Don't save to all three buckets in the database, rather move evidence from one bucket to the next. In addition only two buckets are currently needed: broadcasted and committed. 
- [x] Restore evidenceList at startup if the database contains evidence marked for broadcasting. This is in case of node failure, it can recover evidence that needs to be broadcasted.
- [x] ~Use prioritization algorithm when calling `PendingEvidence` so that evidence can be sent for block proposal in the order of importance. Potentially a weighted average of the voting power and the time to expiry~ I have defaulted with simply using time to order evidence i.e. FIFO - However, I think that it may be best to divide pending evidence between fork causing and non-fork causing in order to dictate priority
- [ ] Ensure that only one evidence per validator is stored and therefore sent to be proposed -  #2561 ( replace only by earlier evidence by the same validator) **I am going to delay doing this until after we have sorted out Amnesia Evidence which involves batches of validators**
- [x] Send evidence in batches. An evidence list is used but only a single piece of evidence is sent at a time. `Receive` is already coded to handle batches of evidence at a time. Fill the max message size with as much evidence is possible.
- [ ] ~~If we don't have any evidence still wait for the `broadcastEvidenceIntervalS` before checking again~~ Better just to be constantly checking for new evidence but should definitely find a way to reduce io - I have considered other options and will leave it as a future issue to also be solved after amnesia evidence
- [x] Better database error handling

Furthermore I have:
- [x] Added batch sending of evidence (during gossiping)
- [x] Removed the info struct and now store the (pending) evidence directly 
- [x] Removed returning an error when evidence is submitted twice because this is an expected behavior of honest nodes and therefore should not be an error. Although this will later need to be addressed in order to avoid DDOS attacks

### **THIS NEEDS TO BE MERGED AFTER THE EVIDENCE HANDLING**

______

For contributor use:

- [x] Wrote tests
- [ ] Updated CHANGELOG_PENDING.md
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: evidence/doc.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/*
+Package evidence handles all evidence storage and gossiping from detection to block proposal.
+For the different types of evidence refer to the `evidence.go` file in the types package.
```
Comment (cmwaters) [inline evidence/doc.go:3]: Do you think that this document should also go through all the evidence types and explain the detection and verification process as well @melekes?

File: evidence/doc.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+go#ConflictingHeadersEvidence)
+ 2. checks that it does not already have the evidence stored
+ 3. verifies the evidence against the nodes state (see state/validation.go#VerifyEvidence)
+ 4. stores the evidence to a db and a concurrent list
+
+The gossiping of evidence is initiated when a peer is added which starts a go routine to broadcast currently
+uncommitted evidence at intervals of 60 seconds (set by the by broadcastEvidenceIntervalS).
+It uses a concurrent list to store the evidence and before sending verifies that each evidence is still valid in the
+sense that it has not exceeded the max evidence age and height which should be set to be equal to the "trusting
+period" (see types/params.go#EvidenceParams).
```
Comment (cmwaters) [inline evidence/doc.go:21]: Do you think we need more information regarding the setting of max evidence age and height (in regards to : https://github.com/tendermint/tendermint/issues/4670) @melekes

File: evidence/doc.go (hunk 1 above)
Comment (melekes) [reply evidence/doc.go:3]: probably not

File: evidence/doc.go (hunk 2 above)
Comment (melekes) [reply evidence/doc.go:21]: no, it should be in ABCI / consensus params documentation, not here.

File: evidence/doc.go (hunk 1 above)
Comment (tessr) [reply evidence/doc.go:3]: Perhaps a link would be helpful, though.

Comment (cmwaters) [reply evidence/doc.go:3]: I will link it to the fork accountability spec at the moment, but in the future, once the evidence spec is written which explains a better overview of the entire evidence process I think we should reference that.

**Comment by cmwaters** [comment]
This PR is big enough that it should really be split into multiple, so I am going to close this in favor of having separate PR's for each of the changes

**Pull request closed without merging**